metricbeat/module/windows/services/services_windows.go

```diff
@@ -0,0 +1,370 @@
+//sys _CloseServiceHandle(handle uintptr) (err error) = advapi32.CloseServiceHandle
+
+var (
+	sizeOfEnumServiceStatusProcess = (int)(unsafe.Sizeof(EnumServiceStatusProcess{}))
+	sizeOfQueryServiceConfig       = (int)(unsafe.Sizeof(QueryServiceConfig{}))
+)
+
+type enumServiceStatusProcess struct {
+	LpServiceName        uintptr
+	LpDisplayName        uintptr
+	ServiceStatusProcess ServiceStatusProcess
+}
+
+type ServiceDatabaseHandle uintptr
+
+type ServiceHandle uintptr
+
+type ProcessHandle uintptr
+
+var serviceStates = map[ServiceState]string{
+	ServiceContinuePending: "ServiceContinuePending",
+	ServicePausePending:    "ServicePausePending",
+	ServicePaused:          "ServicePaused",
+	ServiceRuning:          "ServiceRuning",
+	ServiceStartPending:    "ServiceStartPending",
+	ServiceStopPending:     "ServiceStopPending",
+	ServiceStopped:         "ServiceStopped",
+}
+
+var serviceStartTypes = map[ServiceStartType]string{
+	ServiceAutoStart:   "ServiceAutoStart",
+	ServiceBootStart:   "ServiceBootStart",
+	ServiceDemandStart: "ServiceDemandStart",
+	ServiceDisabled:    "ServiceDisabled",
+	ServiceSystemStart: "ServiceSystemStart",
+}
+
+func (state ServiceState) String() string {
+	if val, ok := serviceStates[state]; ok {
+		return val
+	}
+	return ""
+}
+
+type ServiceStatus struct {
+	DisplayName  string
+	ServiceName  string
+	CurrentState string
+	StartType    string
+}
+
+type ServiceReader struct {
+	handle ServiceDatabaseHandle
+	state  ServiceEnumState
+}
+
+var InvalidServiceDatabaseHandle = ^ServiceDatabaseHandle(0)
+var InvalidServiceHandle = ^ServiceHandle(0)
+
+func OpenSCManager(machineName string, databaseName string, desiredAccess ServiceSCMAccessRight) (ServiceDatabaseHandle, error) {
+	var machineNamePtr *uint16
+	if machineName != "" {
+		var err error
+		machineNamePtr, err = syscall.UTF16PtrFromString(machineName)
+		if err != nil {
+			return InvalidServiceDatabaseHandle, err
+		}
+	}
+
+	var databaseNamePtr *uint16
+	if databaseName != "" {
+		var err error
+		databaseNamePtr, err = syscall.UTF16PtrFromString(databaseName)
+		if err != nil {
+			return InvalidServiceDatabaseHandle, err
+		}
+	}
+
+	handle, err := _OpenSCManager(machineNamePtr, databaseNamePtr, desiredAccess)
+	if err != nil {
+		return InvalidServiceDatabaseHandle, ServiceErrno(err.(syscall.Errno))
+	}
+
+	return handle, nil
+}
+
+func OpenService(handle ServiceDatabaseHandle, serviceName string, desiredAccess ServiceAccessRight) (ServiceHandle, error) {
+	var serviceNamePtr *uint16
+	if serviceName != "" {
+		var err error
+		serviceNamePtr, err = syscall.UTF16PtrFromString(serviceName)
+		if err != nil {
+			return InvalidServiceHandle, err
+		}
+	}
+
+	serviceHandle, err := _OpenService(handle, serviceNamePtr, desiredAccess)
+	if err != nil {
+		return InvalidServiceHandle, ServiceErrno(err.(syscall.Errno))
+	}
+
+	return serviceHandle, nil
+}
+
+func getServiceStates(handle ServiceDatabaseHandle, state ServiceEnumState) ([]ServiceStatus, error) {
```

Comment:
This function is quite long. I recommend breaking some of the pieces into smaller functions then calling those from this function.

Reply:
It's definitely better now, but it still scores a 13 by [gocyclo](https://github.com/fzipp/gocyclo) (<= 10 is what I would aim for). Can you see if you can do a bit more refactoring to this method. Thanks